pkg/mirror-producer/producer.go

```diff
@@ -0,0 +1,224 @@
+// Copyright 2018-2019 Red Hat, Inc.
+// Copyright 2014 CNI authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Go version 1.10 or greater is required. Before that, switching namespaces in
+// long running processes in go did not work in a reliable way.
+//go:build go1.10
+// +build go1.10
+
+package plugin
+
+import (
+	"errors"
+	"fmt"
+	"log"
+	"runtime"
+
+	"github.com/k8snetworkplumbingwg/ovs-cni/pkg/config"
+	"github.com/k8snetworkplumbingwg/ovs-cni/pkg/utils"
+
+	"github.com/containernetworking/cni/pkg/skel"
+	cnitypes "github.com/containernetworking/cni/pkg/types"
+	current "github.com/containernetworking/cni/pkg/types/100"
+
+	"github.com/k8snetworkplumbingwg/ovs-cni/pkg/ovsdb"
+	"github.com/k8snetworkplumbingwg/ovs-cni/pkg/types"
+)
+
+func init() {
+	// this ensures that main runs only on main thread (thread group leader).
+	// since namespace ops (unshare, setns) are done for a single thread, we
+	// must ensure that the goroutine does not jump from OS thread to thread
+	runtime.LockOSThread()
+}
+
+func logCall(command string, args *skel.CmdArgs) {
+	log.Printf("CNI %s was called for container ID: %s, network namespace %s, interface name %s, configuration: %s",
+		command, args.ContainerID, args.Netns, args.IfName, string(args.StdinData[:]))
+}
+
+func getPortUUID(ovsDriver *ovsdb.OvsBridgeDriver, interfaces []*current.Interface) (string, error) {
+	for _, iface := range interfaces {
+		uuid, err := ovsDriver.GetPortUUID(iface.Name)
+		if err == nil {
+			return uuid.GoUUID, nil
+		}
+	}
+
+	return "", errors.New("cannot find port in db")
+}
+
+func attachPortToMirror(ovsDriver *ovsdb.OvsBridgeDriver, portUUIDStr string, mirror *types.Mirror) error {
+	err := ovsDriver.AttachPortToMirrorProducer(portUUIDStr, mirror.Name, mirror.Ingress, mirror.Egress)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func detachPortFromMirror(ovsDriver *ovsdb.OvsBridgeDriver, portUUIDStr string, mirror *types.Mirror) error {
+	err := ovsDriver.DetachPortFromMirrorProducer(portUUIDStr, mirror.Name)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// CmdAdd add handler for attaching container into network
+func CmdAdd(args *skel.CmdArgs) error {
+	logCall("ADD", args)
+
+	netconf, err := config.LoadMirrorConf(args.StdinData)
+	if err != nil {
+		return err
+	}
+
+	ovsDriver, err := ovsdb.NewOvsBridgeDriver(netconf.BrName, netconf.SocketFile)
+	if err != nil {
+		return err
+	}
+
```

Comment:
What happens with orphaned mirrors? Say we called ADD multiple times and then rebooted the machine. Orphaned ports would be cleaned up by the main CNI. Will mirrors be cleaned up too or do we need to clean up from here?

Reply:
I mean ports in OVS DB that were created for veth interface connecting containers. After restart, these ephemeral interfaces are gone, byt the OVS DB entries are still there. These entries will be deleted on the first call of `ovs` ADD. When all these are removed, it could mean that a previously existing mirror does not have any producers anymore and should be cleaned up.

Let me try to come up with a sequence:
1. container1 was attached to br1 through port1, producing to mirror1
2. container2 was attached to br2 through port2, producing to mirror2
3. host is powered off and on
4. host comes up, OVS DB has all the previously created ports and mirrors, but interfaces are not found in the system
5. container1 is scheduled back on this node, container2 is not
6. container1 calls `ovs` CNI ADD, this call runs a cleanup of orphaned port1 and port2, creating a new port3, producing to mirror1

Now mirror 2 is unused but still in the DB. Should `ovs-mirror-producer` CNI call cleanup on ADD, so container1 attachment would clean up mirror2?

Reply:
Ok, now it's clear.
Ideally speaking yes, with a host shutdown this could be the right behaviour, however what will happen in case of a restart of a pod? This will call ADD and it could trigger that behaviour also in a normal scenario.

In ovsdb it's possibile to create a mirror without both producer and consumer and this is exactly what we do in plugins, because both plugins (producer and consumer) try to add a new mirror if not already existing. Next, they try to configure that mirror adding select_* and output_port.
This worked very well in case of multiple pods started at the same time, because concurrent operation are handled very easely.
This means that in a specific moment a mirror could exist without ports. Adding a logic to remove mirrors without ports could cause an unexpected behaviour.

How could ovs-mirror-producer identify mirrors to remove, if ports are cleaned-up by ovs-cni?? Because, It can find all mirrors without ports and remove them. However, how do it know if those are mirrors with errors or not? It's not like the relationship of interfaces and ports. In that case, you can search for interfaces with errors/unexisting, then search related ports and clean up everything.

Also, mirrors could be used to have SPAN ports for SIEM or other security software. It could be dangerous to remove a mirror configured in a plugin from another one. Let me explain:
1) ovs-mirror-* plugin creates mirror 1
2) ovs-mirror-* plugin creates mirror 2
3) shutdown
4) ovs-mirror-* plugin (the one used to create mirror 1)  see the mirror 2 is useless and it removes mirror 2
The last step is quite confusing. Why a plugin configured to operate on mirror-1 is removing mirror-2?
With interfaces its quite easy to find the error condition, but with mirrors it's not so easy. 
We think that ovs-mirror plugin should interact with `Mirror table` only on the specific row (identified by the mirror name passed via NAD). Mirrors are used to read network traffic and a mistake could cause security issues.


But there is another case where the `Mirror` table could become inconsistent:
if you remove a port (either via `cleanPorts` or manually), its UUID reference won't be removed from `Mirror` table (`output` and `select_*` columns).
There are different solutions:
1) ignore the problem
2) when you remove a port via `cleanPorts`, you also need to search and remove it in all mirrors in `Mirror` table, but it means that `ovs-cni` will interact with `Mirror` table even if it shouldn't be its duty
3) when both `ovs-mirror-producer` and `ovs-mirror-consumer` add or delete a port UUID they should also check the existance of all UUIDs referenced in the `Mirror` table for that specific mirror (and not other mirrors, for the same security reason described above)

Reply:
Hi @phoracek do you have any comments about this?

Reply:
Hey, sorry about the delay, I was on a vacation last week. I will re-review this this week

Reply:
> This means that in a specific moment a mirror could exist without ports. Adding a logic to remove mirrors without ports could cause an unexpected behaviour.

Would it be possible create the mirror and attach the first port it in a single transaction? That may fix this issue with transiently unused mirror.

> How could ovs-mirror-producer identify mirrors to remove, if ports are cleaned-up by ovs-cni?? Because, It can find all mirrors without ports and remove them.

I would limit the condition of ovs-mirror-producer/consumer to "are there any ports attached to the mirror". If there are none, we can delete the mirror?

> Why a plugin configured to operate on mirror-1 is removing mirror-2?

It is confusing, but since we don't have any central daemon that could oversee all objects we created, we don't have another choice. It would be consistent with the cleanup done in ovs plugin itself - when invoked to add a port, it would also remove all orphaned ports connected to other bridges it has nothing to do with.

> We think that ovs-mirror plugin should interact with Mirror table only on the specific row (identified by the mirror name passed via NAD).

I see. But this would not allow us to clean up after Pods that after restart got scheduled to a different node, no? Is it possible to mark mirrors managed by the CNI, so we don't touch what's not ours (similar to port cleanup). We created those mirrors, so it should be ok if we remove them - or would it cause issues integrating with other software.

> when both ovs-mirror-producer and ovs-mirror-consumer add or delete a port UUID they should also check the existance of all UUIDs referenced in the Mirror table for that specific mirror (and not other mirrors, for the same security reason described above)

I would be in favor of this one.

---

In general I'm ok with the PR as is. I won't press for the cleanup to be resolved here, since it does not affect the rest of OVS CNI. I am happy to discuss it further though, your call o/

Reply:
Hi @phoracek 

> Would it be possible create the mirror and attach the first port it in a single transaction? That may fix this issue with transiently unused mirror.

In practice it seems feasible, however we need to do some experiments with transaction.


> It is confusing, but since we don't have any central daemon that could oversee all objects we created, we don't have another choice. It would be consistent with the cleanup done in ovs plugin itself - when invoked to add a port, it would also remove all orphaned ports connected to other bridges it has nothing to do with.

It seems that cleanPorts can remove only ports of the current Bridge, even if it finds all interfaces of all bridges with an error (https://github.com/k8snetworkplumbingwg/ovs-cni/blob/e604d234b7895b1b88c89b466e68c56a7a187b54/pkg/ovsdb/ovsdb.go#L190) 
In our opinion this current behaviour seems ok.
If it's possible to create mirrors with a single transaction, we can also add a `cleanMirrors` function with the same logic of `cleanPorts` (limited to the current bridge).


> 5. container1 is scheduled back on this node, container2 is not
> 6. container1 calls ovs CNI ADD, this call runs a cleanup of orphaned port1 and port2, creating a new port3, producing to mirror1

Following the points above, the scenario described in steps 5 and 6 won't cleanup db for both mirrors and ports until some pods are scheduled on the same node with the same bridge and operating on a mirror (for the mirror case)


> I see. But this would not allow us to clean up after Pods that after restart got scheduled to a different node, no? Is it possible to mark mirrors managed by the CNI, so we don't touch what's not ours (similar to port cleanup). We created those mirrors, so it should be ok if we remove them - or would it cause issues integrating with other software.

At the moment, it doesn't seem a big issue and we can work on it in future improvements if required.


> I would be in favor of this one.

We made some tests and this is not required, because if a port is deleted either manually or via cleanPorts, all references to this port will be automatically cleaned across the whole db.

Reply:
Hi @phoracek

> In practice it seems feasible, however we need to do some experiments with transaction.

I'm so stupid. I totally forgot that it's already implemented with transactions :) (https://github.com/auroralabs-eu/ovs-cni/blob/99b6a1b12450175a96d23a5eaeab93504c6a0134/pkg/ovsdb/ovsdb.go#L351)
So, yes, no empty mirrors will be added by the plugin.


> At the moment, it doesn't seem a big issue and we can work on it in future improvements if required.

It's not a big issue, but to be more consistent with ovs-cni we decided to add owner to Mirror table as external_ids.


We pushed some commits with these changes:
- add owner in `external_ids` of Mirror as for Ports. We used the same logic. If a `cmdDel` want to delete a Mirror not owned by ovs-cni, it will return an error.
- implement `cleanMirrors` executed in both `cmdAdd` and `cmdDel` as for `cleanPorts` in `plugin`. Furthermore, as `cleanPorts`, `cleanMirrors` can remove mirrors only if they are in the current bridge and are owned by ovs-cni.
- add some unit tests to cover cases managed by cleanMirrors function:
1. there are some empty mirrors in db owned by ovs-cni and you want to add another mirror to the same bridge -> cleanup empty mirrors
2. there are some empty mirrors in db owned by ovs-cni and you want to remove a mirror of the same bridge -> cleanup empty mirrors
3. there are some empty mirrors in db NOT owned by ovs-cni and you want to add another mirror to the same bridge -> do nothing
4. there are some empty mirrors in db NOT owned by ovs-cni and you want to remove a mirror of the same bridge -> do nothing

We added both add and delete cases to cover cleanMirrors invocation from both `cmdAdd` and `cmdDel`

Reply:
Hey, this is awesome, thanks for adding the cleanup.